File: scripts/ingest-nfl-history.py
#!/usr/bin/env python3
"""
ingest-nfl-history.py

Scaffold for ingesting historical NFL stats into normalized CSVs that
later feed the Postgres database backing NFL Prop Edge.

Quick start (stub mode — writes empty schema-only CSVs, no deps):
    python scripts/ingest-nfl-history.py --season 2025 --weeks 1-10
    python scripts/ingest-nfl-history.py --season 2025 --weeks 1-10 --dry-run

Outputs (relative to --out, default `data/`):
    data/raw/         Raw frames straight from each source pull
                      (one file per source; populated when the
                      nflreadpy / nfl_data_py calls below are wired in)
    data/processed/   Normalized CSVs ready for the Prisma loader:
                        games.csv
                        player_week_stats.csv
                        team_week_stats.csv
                        snap_counts.csv
                        player_ids.csv

Data source (no API keys required):
    nflverse-data releases on GitHub. Wrappers:
        nflreadpy   — Polars-first, actively maintained
        nfl_data_py — Pandas-based (legacy, but stable)

V1 scope: lower-variance markets only. We do not ingest touchdown
columns; they get dropped at the normalization step below.
"""

from __future__ import annotations

import argparse
import csv
import logging
import sys
from pathlib import Path
from typing import Iterable, Optional


# ---------------------------------------------------------------------------
# nflverse imports — uncomment exactly one when ready to leave stub mode.
# Both libraries pull from the same nflverse-data releases (no API keys).
# ---------------------------------------------------------------------------
#
# import nflreadpy as nfl         # Recommended: Polars-first, current.
# import nfl_data_py as nfl_dp    # Legacy: Pandas-based; still works.
#
# Function map used below:
#   schedules         -> nfl.load_schedules(seasons=[s])
#                        nfl_dp.import_schedules([s])
#   player week stats -> nfl.load_player_stats(seasons=[s])
#                        nfl_dp.import_weekly_data([s])
#   play-by-play      -> nfl.load_pbp(seasons=[s])
#                        nfl_dp.import_pbp_data([s])
#   snap counts       -> nfl.load_snap_counts(seasons=[s])
#                        nfl_dp.import_snap_counts([s])
#   players/IDs       -> nfl.load_players()
#                        nfl_dp.import_ids() / import_rosters([s])
#
# ---------------------------------------------------------------------------


LOG = logging.getLogger("ingest-nfl-history")


# ---------------------------------------------------------------------------
# Normalized output schemas.
#
# These column lists are the contract between this script and the Prisma
# loader. Changing them requires a paired Prisma migration. Keep IDs as
# strings so we can store nflverse ids (e.g. `00-0033873` for Mahomes)
# without losing leading zeros.
# ---------------------------------------------------------------------------

GAMES_COLUMNS = [
    "game_id",          # nflverse game_id, e.g. "2025_11_KC_BUF"
    "season",
    "season_type",      # REG | POST
    "week",
    "kickoff_utc",      # ISO-8601 UTC
    "home_team",        # team abbr (KC, BUF, ...)
    "away_team",
    "home_score",
    "away_score",
    "stadium",
    "roof",             # dome | outdoors | closed | open
    "surface",          # grass | fieldturf | ...
    "spread_line",      # closing spread for home team
    "total_line",
]

# Player-week stats restricted to V1 markets (passing/receiving/rushing
# volume). Touchdown columns are intentionally absent.
PLAYER_WEEK_STATS_COLUMNS = [
    "player_id",                # nflverse gsis_id
    "season",
    "week",
    "season_type",
    "team",                     # player's team abbr that week
    "opponent",                 # opponent abbr that week
    "position",                 # QB | RB | WR | TE | ...
    "passing_attempts",
    "passing_completions",
    "passing_yards",
    "receptions",
    "receiving_yards",
    "targets",
    "rushing_attempts",
    "rushing_yards",
    "snaps_offense",            # joined from snap_counts for convenience
]

# Team-week stats are derived from play-by-play. We pre-aggregate here so
# the model layer doesn't need Polars/Pandas just to know how pass-heavy
# a team has been.
TEAM_WEEK_STATS_COLUMNS = [
    "team",
    "season",
    "week",
    "season_type",
    "opponent",
    "plays_offense",
    "plays_defense",
    "pass_attempts_off",
    "pass_attempts_def",
    "rush_attempts_off",
    "rush_attempts_def",
    "passing_yards_off",
    "passing_yards_def",
    "rushing_yards_off",
    "rushing_yards_def",
    "seconds_per_play_off",
    "score_off",
    "score_def",
]

SNAP_COUNTS_COLUMNS = [
    "player_id",
    "season",
    "week",
    "season_type",
    "team",
    "opponent",
    "offense_snaps",
    "offense_pct",
    "defense_snaps",
    "defense_pct",
    "st_snaps",
    "st_pct",
]

PLAYER_IDS_COLUMNS = [
    "player_id",        # canonical id (nflverse gsis_id)
    "gsis_id",
    "esb_id",
    "nflverse_id",
    "pfr_id",
    "sleeper_id",
    "espn_id",
    "full_name",
    "first_name",
    "last_name",
    "position",
    "current_team",
    "birth_date",
    "jersey",
]


# ---------------------------------------------------------------------------
# I/O helpers
# ---------------------------------------------------------------------------

def ensure_dirs(out_root: Path) -> tuple[Path, Path]:
    raw = out_root / "raw"
    processed = out_root / "processed"
    raw.mkdir(parents=True, exist_ok=True)
    processed.mkdir(parents=True, exist_ok=True)
    return raw, processed


def write_csv(path: Path, columns: list[str], rows: Iterable[dict]) -> int:
    n = 0
    with path.open("w", newline="") as fh:
        writer = csv.DictWriter(fh, fieldnames=columns)
        writer.writeheader()
        for row in rows:
            writer.writerow({col: row.get(col, "") for col in columns})
            n += 1
    return n


# ---------------------------------------------------------------------------
# Per-source pulls.
#
# Each function returns an iterable of dicts already mapped to the
# normalized column list above. For now every function is a stub that
# returns an empty iterable — running the script produces schema-only
# CSVs so the downstream loader can be built and tested against the
# expected shape.
#
# When wiring the real pulls, the only thing that changes inside each
# function is the body. The signature stays the same so callers don't
# need to care which library backs the call.
# ---------------------------------------------------------------------------

def pull_schedules(
    season: int, weeks: Optional[set[int]], raw_dir: Path
) -> Iterable[dict]:
    """
    Real implementation (sketch):

        df = nfl.load_schedules(seasons=[season])         # Polars / Pandas
        if weeks:
            df = df.filter(pl.col("week").is_in(list(weeks)))
        df.write_csv(raw_dir / f"schedules-{season}.csv")
        for row in df.iter_rows(named=True):
            yield {
                "game_id":     row["game_id"],
                "season":      row["season"],
                "season_type": row["season_type"],
                "week":        row["week"],
                "kickoff_utc": _to_utc(row["gameday"], row.get("gametime")),
                "home_team":   row["home_team"],
                "away_team":   row["away_team"],
                "home_score":  row.get("home_score"),
                "away_score":  row.get("away_score"),
                "stadium":     row.get("stadium"),
                "roof":        row.get("roof"),
                "surface":     row.get("surface"),
                "spread_line": row.get("spread_line"),
                "total_line":  row.get("total_line"),
            }
    """
    LOG.info("schedules: stub (would pull season=%s weeks=%s)", season, weeks)
    return iter(())


def pull_player_week_stats(
    season: int, weeks: Optional[set[int]], raw_dir: Path
) -> Iterable[dict]:
    """
    Real implementation (sketch):

        df = nfl.load_player_stats(seasons=[season])
        # Source columns of interest:
        #   player_id, season, week, season_type, recent_team, opponent_team,
        #   position, attempts, completions, passing_yards,
        #   receptions, receiving_yards, targets, carries, rushing_yards
        # (We drop *_tds and any TD columns — V1 doesn't trade TDs.)

        if weeks:
            df = df.filter(pl.col("week").is_in(list(weeks)))

        df.write_csv(raw_dir / f"player_stats-{season}.csv")

        for row in df.iter_rows(named=True):
            yield {
                "player_id":          row["player_id"],
                "season":             row["season"],
                "week":               row["week"],
                "season_type":        row["season_type"],
                "team":               row["recent_team"],
                "opponent":           row["opponent_team"],
                "position":           row.get("position"),
                "passing_attempts":   row.get("attempts", 0) or 0,
                "passing_completions":row.get("completions", 0) or 0,
                "passing_yards":      row.get("passing_yards", 0) or 0,
                "receptions":         row.get("receptions", 0) or 0,
                "receiving_yards":    row.get("receiving_yards", 0) or 0,
                "targets":            row.get("targets", 0) or 0,
                "rushing_attempts":   row.get("carries", 0) or 0,
                "rushing_yards":      row.get("rushing_yards", 0) or 0,
                "snaps_offense":      None,  # joined from snap_counts later
            }
    """
    LOG.info(
        "player_week_stats: stub (season=%s weeks=%s)", season, weeks,
    )
    return iter(())


def pull_team_week_stats(
    season: int, weeks: Optional[set[int]], raw_dir: Path
) -> Iterable[dict]:
    """
    Derived from PBP. The PBP frame is ~50k rows per week so we aggregate
    here and only persist the rolled-up totals.

    Real implementation (sketch):

        pbp = nfl.load_pbp(seasons=[season])
        pbp.write_parquet(raw_dir / f"pbp-{season}.parquet")  # parquet, not csv

        if weeks:
            pbp = pbp.filter(pl.col("week").is_in(list(weeks)))

        # Group by (posteam, season, week) for offense and (defteam, ...)
        # for defense, then join the two on the (team, season, week) key.
        # Yield one dict per (team, season, week) matching TEAM_WEEK_STATS_COLUMNS.
    """
    LOG.info("team_week_stats: stub (season=%s weeks=%s)", season, weeks)
    return iter(())


def pull_snap_counts(
    season: int, weeks: Optional[set[int]], raw_dir: Path
) -> Iterable[dict]:
    """
    Real implementation (sketch):

        df = nfl.load_snap_counts(seasons=[season])
        if weeks:
            df = df.filter(pl.col("week").is_in(list(weeks)))
        df.write_csv(raw_dir / f"snap_counts-{season}.csv")
        for row in df.iter_rows(named=True):
            yield {
                "player_id":     row["pfr_player_id"],  # NB: pfr id, may need a join to gsis
                "season":        row["season"],
                "week":          row["week"],
                "season_type":   row["game_type"],
                "team":          row["team"],
                "opponent":      row["opponent"],
                "offense_snaps": row.get("offense_snaps", 0) or 0,
                "offense_pct":   row.get("offense_pct", 0) or 0,
                "defense_snaps": row.get("defense_snaps", 0) or 0,
                "defense_pct":   row.get("defense_pct", 0) or 0,
                "st_snaps":      row.get("st_snaps", 0) or 0,
                "st_pct":        row.get("st_pct", 0) or 0,
            }
    """
    LOG.info("snap_counts: stub (season=%s weeks=%s)", season, weeks)
    return iter(())


def pull_player_ids(season: int, raw_dir: Path) -> Iterable[dict]:
    """
    Real implementation (sketch):

        players = nfl.load_players()                # canonical IDs + bio
        roster  = nfl.load_rosters(seasons=[season])  # current_team / jersey
        # Left-join on gsis_id, then yield one row per player matching
        # PLAYER_IDS_COLUMNS. Cross-id mapping (esb / pfr / sleeper / espn)
        # is the value-add of this file — every other table joins on
        # `player_id` and we resolve other IDs here.
    """
    LOG.info("player_ids: stub (season=%s)", season)
    return iter(())


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def parse_weeks(spec: Optional[str]) -> Optional[set[int]]:
    if not spec:
        return None
    weeks: set[int] = set()
    for part in spec.split(","):
        part = part.strip()
        if not part:
            continue
        if "-" in part:
            lo, hi = part.split("-", 1)
            weeks.update(range(int(lo), int(hi) + 1))
        else:
            weeks.add(int(part))
    return weeks


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Ingest NFL historical stats into normalized CSVs.",
    )
    parser.add_argument(
        "--season", type=int, required=True, help="NFL season year, e.g. 2025",
    )
    parser.add_argument(
        "--weeks",
        help="comma/dash-separated week filter, e.g. '1-10,12'. Default: all weeks.",
    )
    parser.add_argument(
        "--out",
        default="data",
        help="root directory for raw/ and processed/ (default: data)",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Skip writes to disk and just log what would be pulled.",
    )
    args = parser.parse_args(argv)

    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s %(levelname)s %(name)s: %(message)s",
    )

    weeks = parse_weeks(args.weeks)
    out_root = Path(args.out).resolve()
    raw_dir, processed_dir = ensure_dirs(out_root)
    LOG.info("Season=%s weeks=%s", args.season, sorted(weeks) if weeks else "ALL")
    LOG.info("Output: raw=%s processed=%s", raw_dir, processed_dir)

    pulls = [
        ("games",             GAMES_COLUMNS,             pull_schedules(args.season, weeks, raw_dir)),
        ("player_week_stats", PLAYER_WEEK_STATS_COLUMNS, pull_player_week_stats(args.season, weeks, raw_dir)),
        ("team_week_stats",   TEAM_WEEK_STATS_COLUMNS,   pull_team_week_stats(args.season, weeks, raw_dir)),
        ("snap_counts",       SNAP_COUNTS_COLUMNS,       pull_snap_counts(args.season, weeks, raw_dir)),
        ("player_ids",        PLAYER_IDS_COLUMNS,        pull_player_ids(args.season, raw_dir)),
    ]

    if args.dry_run:
        for name, cols, _rows in pulls:
            LOG.info("[dry-run] would write %s (%d cols)", processed_dir / f"{name}.csv", len(cols))
        return 0

    for name, cols, rows in pulls:
        path = processed_dir / f"{name}.csv"
        n = write_csv(path, cols, rows)
        LOG.info("Wrote %s (%d data rows, %d cols)", path, n, len(cols))

    LOG.info("Done.")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: scripts/ingest-nfl-history.ts
/**
 * ingest-nfl-history.ts
 *
 * TypeScript companion to the existing Python scaffold. Reads raw
 * nflverse CSVs from `data/raw/nfl/<season>/` (or fetches them in
 * opt-in network mode), normalizes the rows, and writes the
 * processed CSV bundle to `data/processed/nfl/`.
 *
 * Defaults to `--source local --dry-run`. Network fetch is opt-in
 * via `--source nflverse` AND `ALLOW_NFLVERSE_NETWORK_FETCH=true`.
 *
 *   npx tsx scripts/ingest-nfl-history.ts --season 2025
 *   npx tsx scripts/ingest-nfl-history.ts --start-season 2022 --end-season 2025
 *   npx tsx scripts/ingest-nfl-history.ts --season 2025 --source local
 *   npx tsx scripts/ingest-nfl-history.ts --season 2025 --source nflverse --dry-run
 *
 * No paid APIs. No touchdown columns. No automated betting.
 */

import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import {
  buildNflverseDownloadPlan,
  isNetworkFetchAllowed,
  loadAndNormalizeRaw,
  NETWORK_FETCH_ENV_FLAG,
  writeProcessed,
  type IngestionOptions,
  type NflverseSource,
} from "../src/lib/ingestion/nflverse";

interface CliArgs {
  seasons: number[];
  source: NflverseSource;
  dryRun: boolean;
  rawDir: string;
  outputDir: string;
}

function rangeInclusive(lo: number, hi: number): number[] {
  if (lo > hi) return [];
  const out: number[] = [];
  for (let i = lo; i <= hi; i++) out.push(i);
  return out;
}

function parseArgs(argv: string[]): CliArgs {
  const args: CliArgs = {
    seasons: [],
    source: "local",
    dryRun: true,
    rawDir: path.join(process.cwd(), "data", "raw", "nfl"),
    outputDir: path.join(process.cwd(), "data", "processed", "nfl"),
  };
  let startSeason: number | undefined;
  let endSeason: number | undefined;
  let seasonOverridden = false;
  let dryRunExplicit = false;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const next = () => {
      const v = argv[++i];
      if (v === undefined) throw new Error(`Missing value for ${a}`);
      return v;
    };
    switch (a) {
      case "--season":
        args.seasons = [Number(next())];
        seasonOverridden = true;
        break;
      case "--seasons": {
        const value = next();
        args.seasons = value.split(",").map((v) => Number(v.trim()));
        seasonOverridden = true;
        break;
      }
      case "--start-season":
        startSeason = Number(next());
        break;
      case "--end-season":
        endSeason = Number(next());
        break;
      case "--source": {
        const v = next();
        if (v !== "local" && v !== "nflverse") {
          throw new Error(`Unknown --source value: ${v}`);
        }
        args.source = v;
        break;
      }
      case "--dry-run":
        args.dryRun = true;
        dryRunExplicit = true;
        break;
      case "--no-dry-run":
        args.dryRun = false;
        dryRunExplicit = true;
        break;
      case "--raw":
      case "--raw-dir":
        args.rawDir = next();
        break;
      case "--output":
      case "--out":
      case "--output-dir":
        args.outputDir = next();
        break;
      case "--help":
      case "-h":
        printHelp();
        process.exit(0);
      default:
        throw new Error(`Unknown argument: ${a}`);
    }
  }
  if (!seasonOverridden && (startSeason !== undefined || endSeason !== undefined)) {
    const lo = startSeason ?? endSeason!;
    const hi = endSeason ?? startSeason!;
    args.seasons = rangeInclusive(lo, hi);
  }
  if (args.seasons.length === 0) args.seasons = [2025];

  // Network mode requires explicit env opt-in unless dry-run.
  if (args.source === "nflverse" && !args.dryRun && !isNetworkFetchAllowed()) {
    // Don't silently switch behaviour; surface the gate.
    if (!dryRunExplicit) args.dryRun = true;
  }
  return args;
}

function printHelp(): void {
  // eslint-disable-next-line no-console
  console.log(`Usage:
  npx tsx scripts/ingest-nfl-history.ts [options]

Options:
  --season N              ingest a single season (default 2025)
  --seasons 2024,2025     ingest multiple specific seasons
  --start-season N        inclusive lower bound (range mode)
  --end-season N          inclusive upper bound (range mode)
  --source local          read raw CSVs from --raw (default)
  --source nflverse       fetch from public nflverse-data releases
                          (requires --no-dry-run AND
                           ${NETWORK_FETCH_ENV_FLAG}=true)
  --dry-run               default. Print the plan, write nothing.
  --no-dry-run            actually write processed CSVs
  --raw DIR               raw CSV directory (default data/raw/nfl)
  --output DIR            processed CSV directory
                          (default data/processed/nfl)

Where to put raw nflverse CSVs (one folder per season):
  data/raw/nfl/{season}/schedules.csv
  data/raw/nfl/{season}/player_stats.csv
  data/raw/nfl/{season}/team_stats.csv     (optional)
  data/raw/nfl/{season}/rosters.csv
  data/raw/nfl/{season}/snap_counts.csv    (optional)

Public nflverse-data release URLs (no API key required):
  https://github.com/nflverse/nflverse-data/releases
    /download/schedules/sched_{season}.csv
    /download/player_stats/player_stats_{season}.csv
    /download/rosters/roster_{season}.csv
    /download/snap_counts/snap_counts_{season}.csv

We do NOT ingest touchdown columns. They are dropped at parse time.
`);
}

async function main(): Promise<number> {
  let args: CliArgs;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error(`error: ${(err as Error).message}`);
    printHelp();
    return 2;
  }

  const opts: IngestionOptions = {
    seasons: args.seasons,
    rawDir: args.rawDir,
    processedDir: args.outputDir,
    source: args.source,
    dryRun: args.dryRun,
  };

  // eslint-disable-next-line no-console
  console.log(
    `nflverse ingest — seasons ${args.seasons.join(", ")}, source=${args.source}, dryRun=${args.dryRun}`,
  );

  if (args.source === "nflverse") {
    const plan = buildNflverseDownloadPlan(args.seasons);
    // eslint-disable-next-line no-console
    console.log("Download plan (no network call until --no-dry-run):");
    for (const entry of plan) {
      // eslint-disable-next-line no-console
      console.log(`  Season ${entry.season}:`);
      for (const f of entry.files) {
        // eslint-disable-next-line no-console
        console.log(`    ${f.filename} -> ${f.url}`);
      }
    }
    if (!args.dryRun) {
      if (!isNetworkFetchAllowed()) {
        // eslint-disable-next-line no-console
        console.error(
          `\nerror: network fetch requires ${NETWORK_FETCH_ENV_FLAG}=true. Set the env var and re-run.`,
        );
        return 3;
      }
      // eslint-disable-next-line no-console
      console.error(
        "\nnote: network fetch is scaffolded but not wired in. Drop the files listed above into the --raw directory and re-run with --source local --no-dry-run.",
      );
      return 0;
    }
    // dry-run path returns without writing.
    return 0;
  }

  // local mode — read raw, normalize, optionally write.
  const haveAny = args.seasons.some((s) =>
    fs.existsSync(path.join(opts.rawDir, String(s))),
  );
  if (!haveAny) {
    // eslint-disable-next-line no-console
    console.log(
      `note: no raw season folders found under ${opts.rawDir}. Drop nflverse CSVs there (per --help) and re-run.`,
    );
  }
  const bundle = loadAndNormalizeRaw({
    seasons: args.seasons,
    rawDir: opts.rawDir,
  });
  // eslint-disable-next-line no-console
  console.log(
    `normalized: ${bundle.games.length} games, ${bundle.playerWeekStats.length} player-weeks, ${bundle.teamWeekStats.length} team-weeks, ${bundle.rosters.length} roster entries, ${bundle.snapCounts?.length ?? 0} snap rows`,
  );
  if (args.dryRun) {
    // eslint-disable-next-line no-console
    console.log("dry-run — not writing processed files. Use --no-dry-run to commit.");
    return 0;
  }
  const result = writeProcessed({
    bundle,
    processedDir: opts.processedDir,
  });
  // eslint-disable-next-line no-console
  console.log("written:");
  for (const f of result.written) console.log(`  ${f}`);
  for (const s of result.skipped) console.log(`  skipped: ${s}`);
  return 0;
}

main().then((code) => process.exit(code));
